cmreactor/cmreactor.sh

```diff
@@ -0,0 +1,479 @@
+  else
+    if [[ "$relPath" == "${projectPath}"/* ]]; then
+      relPath="${relPath#"${projectPath}"}"
+    fi
+    relParent="$(dirname "${relPath}")"
+    relName="$(basename "${relPath}")"
+    case "$relParent" in
+    /*) absPath="$(cd "${projectPath}${relParent}" && pwd)/${relName}" ;;
+    *) absPath="$(cd "${projectPath}/${relParent}" && pwd)/${relName}" ;;
+    esac
+    if [[ "${absPath}" == "${projectPath}" ]] || [[ "${absPath}" == "${projectPath}"/* ]]; then
+      echo "${absPath}"
+      return
+    else
+      echo "${absPath} not a child of ${projectPath}" >&2
+      exit 1
+    fi
+  fi
+
+}
+
+reactorPomInit() {
+    local pomFile
+    local module
+    pomFile="$(pathInProject pom.xml)"
+    cat > "${pomFile}" <<EOROOTXMLSTART
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>cmreactor</groupId>
+    <artifactId>cmreactor</artifactId>
+    <version>0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    <modules>
+EOROOTXMLSTART
+while [[ "$#" -gt 0 ]]; do
+    module="$1"
+    shift
+cat >> "${pomFile}" <<EOROOTXMLMOD
+        <module>${module}</module>
+EOROOTXMLMOD
+done
+cat >> "${pomFile}" <<EOROOTXMLEND
+    </modules>
+</project>
+EOROOTXMLEND
+}
+
+readConfigValue() {
+    local config
+    local key
+    local defaultValue
+
+    if [[ $# -lt 2 ]]; then
+        echo "readConfigValue requires two args: config key [optional default]" >&2
+        exit 1
+    fi
+
+    config="$1"
+    key="$2"
+    defaultValue="$3"
+
+    if gitc config -f "${config}" "${key}" >/dev/null; then
+        gitc config -f "${config}" "${key}"
+    elif [[ "$defaultValue" != "" ]]; then
+        echo "$defaultValue"
+    fi
+}
+
+readGitConfigSyncElements() {
+    local config
+    config="$1"
+
+    if gitc config -f "${config}" --get-all "cmreactor.configsync" >/dev/null; then
+        gitc config -f "${config}" --get-all "cmreactor.configsync"
+    else
+        echo user.email
+        echo user.name
+    fi
+}
+
+readModulesFromConfig() {
+    local config
+    config="$1"
+
+    if gitc config -f "${config}" --get-all "cmreactor.modules" >/dev/null; then
+        gitc config -f "${config}" --get-all "cmreactor.modules"
+    fi
+}
+
+readModuleConfigValue() {
+    local config
+    local module
+    local key
+    local defaultValue
+
+    if [[ $# -lt 3 ]]; then
+        echo "readModuleConfigValue requires three args: config module key [optional default]" >&2
+        exit 1
+    fi
+
+    config="$1"
+    module="$2"
+    key="$3"
+    defaultValue="$4"
+
+    if gitc config -f "${config}" "cmreactor.module.${module}.${key}" >/dev/null; then
+        gitc config -f "${config}" "cmreactor.module.${module}.${key}"
+    elif [[ "$defaultValue" != "" ]]; then
+        echo "$defaultValue"
```

Comment:
Is there any value to the `$defaultValue` implementation here which simply echo's it?  Doesn't look like any calls to this function are passing the 4th argument.

Reply:
in a bash function, you have to echo or otherwise print the value to stdout so that the caller can assign it to a variable. `return` statements actually set the exit code of the function call.

Reply:
got it, thanks